packages/subgraph/src/util.ts

```diff
@@ -205,3 +205,97 @@ export function createTransactionHistory(
 
   return transactionHistory;
 }
+
+// buy: amount > 0
+// sell amount < 0
+export function buyOrSellETH(userAddr: string, amount: BigDecimal) {
+  let usdcPrices = getETHUSDCPrice();
+
+  let position = loadOrCreatePosition(userAddr);
+
+  // Buy
+  if (amount.gt(ZERO_BD)) {
+    let oldBoughtAmount = position.currentETHAmount.plus(
+      position.realizedETHAmount
+    );
+    let oldRealizedETHGain =
+      position.realizedETHUnitGain.times(oldBoughtAmount);
+
+    position.realizedETHUnitGain = oldRealizedETHGain
+      .plus(amount.times(usdcPrices[1]))
+      .div(oldBoughtAmount.plus(amount));
+  }
+
+  // Sell
+  if (amount.lt(ZERO_BD)) {
+    let absAmount = amount.neg();
+    let oldRealizedETHCost = position.realizedETHUnitCost.times(
+      position.realizedETHAmount
+    );
+
+    position.realizedETHAmount = position.realizedETHAmount.plus(absAmount);
+    position.realizedETHUnitCost = oldRealizedETHCost
+      .plus(absAmount.times(usdcPrices[1]))
+      .div(position.realizedETHAmount);
+  }
+
+  // Unrealized PnL calculation
+  let oldUnrealizedETHCost = position.unrealizedETHUnitCost.times(
+    position.currentETHAmount
+  );
+
+  position.currentETHAmount = position.currentETHAmount.plus(amount);
+  position.unrealizedETHUnitCost = oldUnrealizedETHCost
+    .plus(amount.times(usdcPrices[1]))
+    .div(position.currentETHAmount);
+
+  position.save();
+}
+
+// buy: amount > 0
+// sell amount < 0
+export function buyOrSellSQTH(userAddr: string, amount: BigDecimal) {
+  let usdcPrices = getETHUSDCPrice();
+  let osqthPrices = getoSQTHETHPrice();
+  let osqthPrice = osqthPrices[1].times(usdcPrices[1]);
+
+  let position = loadOrCreatePosition(userAddr);
+
+  // Buy
+  if (amount.gt(ZERO_BD)) {
+    let oldBoughtAmount = position.currentOSQTHAmount.plus(
+      position.realizedOSQTHAmount
+    );
+    let oldRealizedOSQTHGain =
+      position.realizedOSQTHUnitGain.times(oldBoughtAmount);
+
+    position.realizedOSQTHUnitGain = oldRealizedOSQTHGain
+      .plus(amount.times(osqthPrice))
+      .div(oldBoughtAmount.plus(amount));
+  }
+
+  // Sell
+  if (amount.lt(ZERO_BD)) {
+    let absAmount = amount.neg();
+    let oldRealizedOSQTHCost = position.realizedOSQTHUnitCost.times(
+      position.realizedOSQTHAmount
+    );
+
+    position.realizedOSQTHAmount = position.realizedOSQTHAmount.plus(absAmount);
+    position.realizedOSQTHUnitCost = oldRealizedOSQTHCost
+      .plus(absAmount.times(osqthPrice))
+      .div(position.realizedOSQTHAmount);
+  }
+
```

Comment:
Hey Adam , we should also add position type change here as well, after users buying or selling osqth